OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs

```diff
@@ -90,42 +88,55 @@ public void Tick(Actor self)
 				return;
 
 			VisibilityHash = 0;
-			foreach (var player in self.World.Players)
+			var players = self.World.Players;
+
+			if (!initialized)
+			{
+				// The world players never change, so we can safely index this collection.
+				stateByPlayerIndex = new FrozenState[players.Length];
+				tooltip = self.TraitsImplementing<ITooltip>().FirstOrDefault();
+				health = self.TraitOrDefault<Health>();
+			}
+
+			for (var i = 0; i < players.Length; i++)
 			{
 				FrozenActor frozenActor;
 				bool isVisible;
 				if (!initialized)
 				{
+					var player = players[i];
 					frozenActor = new FrozenActor(self, footprint, player.Shroud, startsRevealed);
 					isVisible = startsRevealed;
-					stateByPlayer.Add(player, new FrozenState(frozenActor) { IsVisible = isVisible });
+					var state = new FrozenState(frozenActor) { IsVisible = isVisible };
+					stateByPlayer.Add(player, state);
+					stateByPlayerIndex[i] = state;
 					player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
 				}
 				else
 				{
-					var state = stateByPlayer[player];
+					var state = stateByPlayerIndex[i];
 					frozenActor = state.FrozenActor;
 					isVisible = !frozenActor.Visible;
 					state.IsVisible = isVisible;
 				}
 
 				if (isVisible)
-					VisibilityHash += player.ClientIndex;
+					VisibilityHash |= 1 << (i % 32);
```

Comment:
It's just a hash. Besides, it's better than adding the raw player indexes which generates far more collisions and leaves the upper bits unused.